Alias OSError to TimeoutError in python < 3.3

cledoux opened the pull request:
Why:

* TimeoutError doesn't appear to exist anywhere.
* It's giving an error when timeouts actually occur.

This change addresses the need by:

* Simply remove the exception entirely.

cledoux commented:
In addition, do you plan to submit a PR to the main neomodel repo?  If so, anything I can do to help?

Your changes are solving a problem I'm currently having in production and it would be nice to have these changes merged with the official repo and published to pypi.

jrouter commented:
The TimeoutError is thrown by the neo4j python driver. It is a built in exception in Python:

https://docs.python.org/3/library/exceptions.html

This error occurs frequently when you are running a load balancer in front of the neo4j cluster and the connection is timed out by the lb. I'm not quite sure why catching this error would have any effects on connection timeouts; if you can add detail regarding this scenario, it would be much appreciated.

As for submitting a pull request, I do plan on doing this with TimeoutError handling, as not handling this error causes the connection pool to be populated with dead connections when running behind an lb. However, I wanted to bake this a little longer to see what effect it has on the atomicity of db.transactions. You do lose the atomicity as the transaction session goes null or there is an exception thrown by neomodel because there is a "transaction in progress".

cyledoux committed:
Alias OSError to TimeoutError in python < 3.3

Why:

* The TimeoutError is sometimes thrown by the neo4j driver in python3.
* This error is being caught in neomodel.util in order to retry the
  connection.
* TimeoutError was introduced in Python 3.3 and so doesn't exist in
  python 2.7

This change addresses the need by:

* If in python < 3.3, alias OSError to TimeoutError.
* TimeoutError is a subclass of OSError, so I expect OSError to be
  raised instead of TimeoutError in python 2.7.

cledoux commented:
`TimeoutError` was introduced in Python 3.3 and we are still (unfortunately) running python 2.7.  Thus, when a `ServiceUnavailable` exception is thrown, python complains about `TimeoutError` not existing.

The line causing the problem is in `neomodel.util.Database.cypher_query` (trimmed to only the relevant pieces below):

``` python
class Database(local):
    ...
    def cypher_query(self, query, params=None, handle_unique=True, retry_on_session_expire=False, retry_failed_connection=0):
       ...
        try:
            start = time.clock()
            response = session.run(query, params)
            results, meta = [list(r.values()) for r in response], response.keys()
            end = time.clock()
        ...
        # When this except block is triggered, python 2.7 complains.
        except (TimeoutError,ServiceUnavailable) as err:
            ...

        return results, meta
```

I've updated my PR to alias OSError to TimeoutError when sys.version < (3,3).  I'm assuming that in python 2.7 the exception will be an OSError instead of a TimeoutError because TimeoutError is a subclass of OSError.  I don't have a load balanced cluster to test this against at the moment.

Glad to know you plan to submit to upstream.  Like I said, this is solving a problem I'm having as well.  Let me know if you need any help. My email is in my github profile.

cledoux commented:
Closing the PR and reopening using a feature branch.  I have some changes I want to push to my master branch without affecting the PR. Apologies if this causes any issues.
